[TFLite] Fix cmake build kernel test fail

bhbruce opened the pull request:
- Problem:

Using CMake to build the TFLite kernel test will fail in linking stage.

- Error log

```
[ 87%] Linking CXX executable unpack_test
ld: libtensorflow-lite-test-base.a(op_version.cc.o): in function `.LBB1_18':
op_version.cc:(.text._ZN6tflite15UpdateOpVersionEPh+0x14c): undefined reference to `tflite::GetOpSignature(tflite::OperatorCode const*, tflite::Operator const*, tflite::SubGraph const*, tflite::Model const*)'
clang++: rror: linker command failed with exit code 1 (use -v to see invocation)
```

- Solution:

Add `versioning/op_signature.cc` to `TEST_FRAMEWORK_SRC`. `versioning/op_version.cc` depends on `versioning/op_signature.cc` because of using GetOpSignature().

bhbruce committed:
[TFLite] Fix cmake build kernel test fail

Using CMake to build the TFLite kernel test will fail in linking stage.

ld: libtensorflow-lite-test-base.a(op_version.cc.o): in function `.LBB1_18':
op_version.cc:(.text._ZN6tflite15UpdateOpVersionEPh+0x14c): undefined reference to `tflite::GetOpSignature(tflite::OperatorCode const*, tflite::Operator const*, tflite::SubGraph const*, tflite::Model const*)'
clang++: rror: linker command failed with exit code 1 (use -v to see invocation)

Solution:
Add `versioning/op_signature.cc` to TEST_FRAMEWORK_SRC.
`versioning/op_version.cc` depends on `versioning/op_signature.cc` because of using GetOpSignature().

Signed-off-by: Bruce Lai <bruce.lai@sifive.com>

bhbruce commented:
@terryheo @NAgarwalla 
Can you help to review it?

gbaned commented:
Hi @grantjensen / @terryheo Can you please review this PR ? Thank you!

grantjensen commented:
Failing internal test when linking conv_test:

```[100%] Linking CXX executable conv_test
/usr/bin/ld: ../libtensorflow-lite.a(op_signature.cc.o): in function `tflite::GetOpSignature(TfLiteContext const*, TfLiteNode const*, TfLiteRegistration const*)':
op_signature.cc:(.text+0x440): multiple definition of `tflite::GetOpSignature(TfLiteContext const*, TfLiteNode const*, TfLiteRegistration const*)'; libtensorflow-lite-test-base.a(op_signature.cc.o):op_signature.cc:(.text+0x440): first defined here
/usr/bin/ld: ../libtensorflow-lite.a(op_signature.cc.o): in function `tflite::GetOpSignature(tflite::OperatorCode const*, tflite::Operator const*, tflite::SubGraph const*, tflite::Model const*)':
op_signature.cc:(.text+0xb60): multiple definition of `tflite::GetOpSignature(tflite::OperatorCode const*, tflite::Operator const*, tflite::SubGraph const*, tflite::Model const*)'; libtensorflow-lite-test-base.a(op_signature.cc.o):op_signature.cc:(.text+0xb60): first defined here
collect2: error: ld returned 1 exit status```

Please fix

bhbruce commented:
> Failing internal test when linking conv_test:
> 
> ```
> /usr/bin/ld: ../libtensorflow-lite.a(op_signature.cc.o): in function `tflite::GetOpSignature(TfLiteContext const*, TfLiteNode const*, TfLiteRegistration const*)':
> op_signature.cc:(.text+0x440): multiple definition of `tflite::GetOpSignature(TfLiteContext const*, TfLiteNode const*, TfLiteRegistration const*)'; libtensorflow-lite-test-base.a(op_signature.cc.o):op_signature.cc:(.text+0x440): first defined here
> /usr/bin/ld: ../libtensorflow-lite.a(op_signature.cc.o): in function `tflite::GetOpSignature(tflite::OperatorCode const*, tflite::Operator const*, tflite::SubGraph const*, tflite::Model const*)':
> op_signature.cc:(.text+0xb60): multiple definition of `tflite::GetOpSignature(tflite::OperatorCode const*, tflite::Operator const*, tflite::SubGraph const*, tflite::Model const*)'; libtensorflow-lite-test-base.a(op_signature.cc.o):op_signature.cc:(.text+0xb60): first defined here
> collect2: error: ld returned 1 exit status```
> 
> Please fix
> ```

Ok, I'll fix it.

bhbruce commented:
> /usr/bin/ld: ../libtensorflow-lite.a(op_signature.cc.o): in function `tflite::GetOpSignature(TfLiteContext const*, TfLiteNode const*, TfLiteRegistration const*)':
> op_signature.cc:(.text+0x440): multiple definition of `tflite::GetOpSignature(TfLiteContext const*, TfLiteNode const*, TfLiteRegistration const*)'; libtensorflow-lite-test-base.a(op_signature.cc.o):op_signature.cc:(.text+0x440): first defined here
> /usr/bin/ld: ../libtensorflow-lite.a(op_signature.cc.o): in function `tflite::GetOpSignature(tflite::OperatorCode const*, tflite::Operator const*, tflite::SubGraph const*, tflite::Model const*)':
> op_signature.cc:(.text+0xb60): multiple definition of `tflite::GetOpSignature(tflite::OperatorCode const*, tflite::Operator const*, tflite::SubGraph const*, tflite::Model const*)'; libtensorflow-lite-test-base.a(op_signature.cc.o):op_signature.cc:(.text+0xb60): first defined here
> collect2: error: ld returned 1 exit status```

@grantjensen Can you share more information about which compiler you used (e.g gcc for arm64)?

grantjensen commented:
Sure.

```
"${CMAKE}" ../tensorflow/lite -DTFLITE_ENABLE_GPU=ON -DTFLITE_KERNEL_TEST=ON
time "${CMAKE}" --build . -j -t benchmark_model -t conv_test
```
(Cmake -v == 3.16.8)
I have also run this code with unpack_test and have not encountered any errors. Do you mind providing the code you used to generate the original issue?

bhbruce commented:
> Do you mind providing the code you used to generate the original issue?

I set `-DTFLITE_ENABLE_GPU=OFF` and the bug happens as I mentioned at the top.

The multiple define errors happen only when `-DTFLITE_ENABLE_GPU=ON`.
Because `op_signature` is included in `TFLITE_DELEGATES_GPU_SRCS`, it makes `tensorflow-lite.a` define `GetOpSignature `.
 https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/CMakeLists.txt#L603C22-L603C22

My previous modification makes `tensorflow-lite-test-base` include the definition of `op_signature`.
https://github.com/tensorflow/tensorflow/blob/37f535948ea6ff280100ab254c7d9acd183e2ed6/tensorflow/lite/kernels/CMakeLists.txt#L122C57-L122C57
`tensorflow-lite-test-base` will later link  `tensorflow-lite`(it defines op_signature when `TFLITE_ENABLE_GPU=ON`) later, so it causes multiple definitions here.
https://github.com/tensorflow/tensorflow/blob/37f535948ea6ff280100ab254c7d9acd183e2ed6/tensorflow/lite/kernels/CMakeLists.txt#L122C13-L127

bhbruce committed:
[TFLite] Avoid multiple def when cmake build w/ GPU

The multiple define errors happen only when `-DTFLITE_ENABLE_GPU=ON`.
Because op_signature is appended in `TFLITE_DELEGATES_GPU_SRCS`, it makes `tensorflow-lite.a` define `GetOpSignature`.

This commit makes `tensorflow-lite-test-base` include the definition of `GetOpSignature` only if `TFLITE_ENABLE_GPU=OFF`.

Signed-off-by: Bruce Lai <bruce.lai@sifive.com>

bhbruce commented:
I provide another [PR](https://github.com/tensorflow/tensorflow/pull/61798) to solve this issue because of my operation error.
